src/protocols/rdp/rdp.c

```diff
@@ -227,10 +234,60 @@ static BOOL rdp_freerdp_authenticate(freerdp* instance, char** username,
 
     rdpContext* context = instance->context;
     guac_client* client = ((rdp_freerdp_context*) context)->client;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    char* params[4] = {};
+    int i = 0;
+    
+    /* If the client does not support the "required" instruction, warn and
+     * quit.
+     */
+    if (!guac_client_owner_supports_required(client)) {
+        guac_client_log(client, GUAC_LOG_WARNING, "Client does not support the "
+                "\"required\" instruction. No authentication parameters will "
+                "be requested.");
+        return TRUE;
+    }
 
-    /* Warn if connection is likely to fail due to lack of credentials */
-    guac_client_log(client, GUAC_LOG_INFO,
-            "Authentication requested but username or password not given");
+    /* If the username is undefined, add it to the requested parameters. */
+    if (settings->username == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_USERNAME, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_USERNAME;
+        i++;
+    }
+    
+    /* If the password is undefined, add it to the requested parameters. */
+    if (settings->password == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_PASSWORD, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_PASSWORD;
+        i++;
+    }
+    
+    /* If the domain is undefined, add it to the requested parameters. */
+    if (settings->domain == NULL) {
+        guac_argv_register(GUAC_RDP_ARGV_DOMAIN, guac_rdp_argv_callback, NULL, 0);
+        params[i] = GUAC_RDP_ARGV_DOMAIN;
+        i++;
+    }
+    
+    /* NULL-terminate the array. */
+    params[i] = NULL;
+    
+    if (i > 0) {
+        
+        /* Send required parameters to the owner. */
+        guac_client_owner_send_required(client, (const char**) params);
+        
+        guac_argv_await((const char**) params);
+        
+        /* Get new values from settings. */
+        *username = settings->username;
+        *password = settings->password;
+        *domain = settings->domain;
```

Comment:
There may already be pointers to allocated memory assigned to `*username`, `*password`, etc. which will orphaned through these assignments. The existing values should first be freed.

As FreeRDP will also automatically free the assigned strings, and we already free our copies of these strings stored within `settings`, these should be duplicated so that they aren't double-freed. This is how things are currently done within `guac_rdp_push_settings()`:

https://github.com/apache/guacamole-server/blob/382d72a26a04008aa936680deed00e3a31086b1e/src/protocols/rdp/settings.c#L1296-L1299

Reply:
As `guac_rdp_strdup()` is only defined within the context of `settings.c`, is using `strdup()` acceptable, here, or do we need to pull `guac_rdp_strdup()` up into a header file and use that?

Reply:
I think `guac_rdp_strdup()` is needed here. The standard `strdup()` will segfault if given `NULL`, whereas `guac_rdp_strdup()` will simply and safely return `NULL`.

The definition for `guac_rdp_strdup()` could be moved to `settings.h`. Or we could consider adding a `guac_strdup()` to libguac's `string.h`.

Reply:
Went ahead and moved it to libguac's `string.h` and rename it `guac_strdup()`.